Replace use of Create React App with Vite

ecmps-ui and campd-ui both currently use Create React App, which has a vulnerable dependency on react-scripts.  

This dependency is outdated and appears to no longer be maintained.  A number of our open Dependabot alerts include react-scripts as a dependency that is blocking resolution.

A number of other ERG-maintained react apps have switched to Vite, and this appears to be a widely accepted build tool.


### Requirements:  
1. Replace use of Create React App with Vite
2. Update testcase with jest.
3. Update uswds css with uswds-compile

#### Step 1: Install Vite and Vite React Plugin
 Install Vite and the Vite React plugin.

```yarn add vite @vitejs/plugin-react --dev ```
#### Step 2: Create a Vite Configuration File
Create a vite.config.js file in the root, with the following content:
```
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

export default defineConfig({
  plugins: [react()],
  server: {
    port: 3000, // or any other port
  },
  build: {
    outDir: 'build', // 
  },
});
```
#### Step 3: Update package.json
Update package.json to use Vite scripts instead of CRA scripts:

```
{
  "scripts": {
    "start": "vite",
    "build": "vite build",
    "serve": "vite preview",
    "test": "jest --coverage --testResultsProcessor=jest-sonar-reporter",
    "test:cov": "jest --coverage",
    "up": "yarn && yarn start"
  }
}
```

#### Step 4: Move index.html
Move index.html file from the public directory to the root. Update the paths to JavaScript and CSS files in the index.html file:
```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="description" content="ECMPS is your..." />
    <meta name="keywords" content="..." />
    <meta property="og:site_name" content="Emissions Collection & Monitoring Plan System (ECMPS)">
    <meta property="og:url" content="https://ecmps.epa.gov">
    <meta name="MobileOptimized" content="width">
    <meta name="HandheldFriendly" content="true">

    <script src="/env-config.js"></script>
    
    <link rel="icon" href="/favicon.ico" />
    <link rel="manifest" href="/manifest.json" />

    <title>Emissions Collection & Monitoring Plan System (ECMPS)</title>
  </head>
  <body>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <div id="root" aria-hidden="false"></div>
    <script type="module" src="/src/index.js"></script>
  </body>
</html>
```
#### Step 5: Update Imports
Vite uses ES modules by default, so might need to update some of imports. For example, if any CommonJS require statements, should convert them to ES module import statements.
```
// const validate = require('jsonschema').validate;
import {validate} from 'jsonschema';
```
#### Step 6: Update Environment Variables
Vite uses a different prefix for environment variables. Replace REACT_APP_ with VITE_ in environment variables. For example, `REACT_APP_EASEY_ECMPS_UI_API_KEY` becomes `VITE_EASEY_ECMPS_UI_API_KEY`.

#### Step 7: Remove CRA Dependencies
``` yarn remove react-scripts craco```

#### Step 8: Update getting environment veriable 
~~`import.meta.env` instead of `process.env`~~
Jest and Vite: Cannot use `import.meta` outside a module [DOC](https://dev.to/rubymuibi/jest-and-vite-cannot-use-importmeta-outside-a-module-24n3)

#### Step 9: Update package.json
 `{"type": "module"}`
update `sass` and `@uswds/uswds` version 

#### Setp 10: Create a Jest Configuration File for testing
```
export default {
  testEnvironment: 'jsdom',
  setupFilesAfterEnv: ['<rootDir>/src/setupTests.js'], 
  moduleNameMapper: {
    '\\.(css|scss)$': '<rootDir>/src/mocks/styleMock.js',
    'react-markdown': '<rootDir>/src/mocks/styleMock.js',
    'remark-gfm': '<rootDir>/src/mocks/styleMock.js',
  },
  transform: {
    '^.+\\.(js|jsx)$': 'babel-jest',
  },
  transformIgnorePatterns: [
    "/node_modules/(?!(axios|other-module)/)",
  ],
  testMatch: ['**/?(*.)+(test).[jt]s?(x)'],
  collectCoverageFrom: [
    'src/**/*.{js,jsx}',
    '!src/index.js',
    '!src/App.js',
    '!src/config.js',
    '!src/serviceWorker.js',
    '!src/store/configureStore.dev.js',
    '!src/store/actions/actionTypes.js',
    '!src/store/reducers/index.js',
    '!src/store/reducers/initialState.js',
    '!src/utils/api/axiosSetup.js',
    '!src/utils/api/setupTests.js',
  ],
  coverageReporters: ['html', 'text', 'clover', 'json', 'lcov'],
};
```

#### Step 11: Create a Babel Configuration File.
Create a .babelrc file in the root of your project
```
{
	"presets": [
	  ["@babel/preset-env", { "targets": { "node": "current" } }],
	  ["@babel/preset-react", { "runtime": "automatic" }] // Enable the new JSX transform
	]
  }
```

#### Step 12: Update uswds css with uswds-compile. 
Vite did not support current method it make erorr
``` Can't find stylesheet to import.
@use [sass] "uswds-core" Can't with find (........ stylesheet to import.
 @use "uswds-core" with (........
```
So update the uswds by using **[@uswds/compile](https://designsystem.digital.gov/documentation/getting-started/developers/phase-two-compile/#using-uswds-compile-2)**
Need to create separate CSS and JS files, use PostCSS to extract the styles. https://github.com/US-EPA-CAMD/easey-design-system/blob/b60dc0d3c9cff83264ba3ab2d3b77702ab55da9b/rollup.config.js#L43
```
postcss({
        extract: true,
        modules: false,
        use: ['sass'],
      }),
```

and export the package.
in package.json
```
"exports": {
    ".": {
      "import": "./index.esm.js",
      "require": "./index.js"
    }
  },
```
~~By using uswds-compile need to update the [devops build.sh](https://github.com/US-EPA-CAMD/devops/blob/master/scripts/yarn/build.sh), for updating~~
 ```
npx gulp init
```

#### After that will get an error 
#### 1 `[ERROR] The JSX syntax extension is not currently enabled`
 it means Vite is not recognizing JSX syntax in your `.js` files. By default, Vite expects JSX to be in `.jsx` or `.tsx` files. However, you can configure Vite to allow JSX in `.js` files as well.
Modify `vite.config.js` file
```
esbuild: {
        loader: "jsx",
        include: /src\/.*\.jsx?$/,
        exclude: [],
    },
    optimizeDeps: {
        esbuildOptions: {
            loader: {
                '.js': 'jsx',
            },
        },
    },
```
#### 2 `Pre-transform error: Failed to resolve entry for package "@us-epa-camd/easey-design-system". The package may have incorrect main/module/exports specified in its package.json.`

This typically happens when the package.json of the package does not have a valid main, module, or exports field, or the specified entry point does not exist.

~~Modify  vite.config.js using the resolve.alias option:~~

~~`import path from 'path';`~~

~~resolve: {~~
        ~~alias: {~~
            ~~'@us-epa-camd/easey-design-system': path.resolve(~~
                ~~__dirname,~~
                ~~'node_modules/@us-epa-camd/easey-design-system/index.js'~~
            ~~),~~
       ~~},~~
   ~~}~~


Add export in easey-design-system package .
```
"exports": {
    ".": {
      "import": "./index.esm.js",
      "require": "./index.js"
    }
  },
```

#### 3. `Deprecation Warning [import]: Sass @import rules are deprecated and will be removed in Dart Sass 3.0.0.`
The warning seeing is because the `@import` rule in Sass has been deprecated, and Sass is moving toward using the new `@use` rule for importing stylesheets. In the future, `@import` will be removed in Dart Sass 3.0.0.

To resolve this, should switch to using `@use` instead of `@import`. 

```
use ---> @use "../../styles/variables" as *; instead of @import "../../styles/variables";

// other update
@use "global-styles" as *;
@use "icons" as *;
@use "transitions" as *;
```

#### 4  crypto.getRandomValues() not supported. 
In test case crypto.getRandomValues() not supported See https://github.com/uuidjs/uuid#getrandomvalues-not-supported
```
npm install react-native-get-random-values
// Add this line to your `index.js`
import 'react-native-get-random-values'
``` 
 OR

```
// Add this in setupTests.js
import nodeCrypto from 'crypto';
window.crypto = {
  getRandomValues: function (buffer) {
    return nodeCrypto.randomFillSync(buffer);
  }
}
```

--- a/src/components/MonitoringPlanTab/MonitoringPlanTab.js
+++ b/src/components/MonitoringPlanTab/MonitoringPlanTab.js
@@ -59,7 +59,7 @@ export const MonitoringPlanTab = ({
   }, [locationSelect]);
 
   return (
-    <div>
+    <div id="monitoring-plan-tab-container">
       <div>
         <MonitoringPlanTabRender
           resetTimer={resetTimer}

--- a/src/components/MonitoringPlanTab/MonitoringPlanTab.scss
+++ b/src/components/MonitoringPlanTab/MonitoringPlanTab.scss
@@ -1,7 +1,9 @@
-.usa-accordion__button {
-  background-color: white;
-  max-width: 100%;
-}
-.usa-accordion__content {
-  padding: 0 !important;
+#monitoring-plan-tab-container {
+	.usa-accordion__button {
+		background-color: white;
+		max-width: 100%;
+	}
+	.usa-accordion__content {
+		padding: 0 !important;
+	}
 }

--- a/src/components/Tabs/Tabs.js
+++ b/src/components/Tabs/Tabs.js
@@ -120,11 +120,7 @@ const Tabs = ({
                     tabIndex="0"
                     id="select-config"
                     aria-label={`open ${el.props.title} tab`}
-                    className={
-                      currentTabIndex === i
-                        ? "initial-tab-button active-tab-button"
-                        : "initial-tab-button"
-                    }
+                    className="initial-tab-button"
                     onClick={() => setCurrentTabIndex(i)}
                   >
                     {el.props.title}

